dali/operators/reader/loader/video/video_file.cc

```diff
@@ -0,0 +1,261 @@
+    entry.is_keyframe = false;
+    entry.pts = av_state_->frame_->pts;
+    entry.is_flush_frame = true;
+    entry.last_keyframe_id = last_keyframe;
+    index_.push_back(entry);
+  }
+  Reset();
+}
+
+void VideoFileCPU::CopyToOutput(uint8_t *data) {
+  dest_[0] = data;
+  dest_linesize_[0] = av_state_->frame_->width * Channels();
+
+  int ret = sws_scale(
+    av_state_->sws_ctx_,
+    av_state_->frame_->data,
+    av_state_->frame_->linesize,
+    0,
+    av_state_->frame_->height,
+    dest_,
+    dest_linesize_);
+
+  DALI_ENFORCE(
+    ret >= 0,
+    make_string("Could not convert frame data to RGB: ", detail::av_error_string(ret)));
+}
+
+bool VideoFileCPU::ReadRegularFrame(uint8_t *data, bool copy_to_output) {
+  int ret = -1;
+  while(av_read_frame(av_state_->ctx_, av_state_->packet_) >= 0) {
+    if (av_state_->packet_->stream_index != av_state_->stream_id_) {
+      continue;
+    }
+
+    ret = avcodec_send_packet(av_state_->codec_ctx_, av_state_->packet_);
+    DALI_ENFORCE(
+      ret >= 0,
+      make_string("Failed to send packet to decoder: ", detail::av_error_string(ret)));
+
+    ret = avcodec_receive_frame(av_state_->codec_ctx_, av_state_->frame_);
+
+    if (ret == AVERROR(EAGAIN) || ret == AVERROR_EOF) {
+      continue;
+    }
+
+    if (!copy_to_output) {
+      return true;
+    }
+
+    if (av_state_->sws_ctx_ == nullptr) {
+      av_state_->sws_ctx_ = sws_getContext(
+        Width(), 
+        Height(), 
+        av_state_->codec_ctx_->pix_fmt, 
+        Width(), 
+        Height(), 
+        AV_PIX_FMT_RGB24, 
+        SWS_BILINEAR, 
+        nullptr, 
+        nullptr, 
+        nullptr);
+      DALI_ENFORCE(av_state_->sws_ctx_, "Could not create sw context");
+    }
+
+    CopyToOutput(data);
+    return true;
+  }
+
+  ret = avcodec_send_packet(av_state_->codec_ctx_, nullptr);
+  DALI_ENFORCE(
+    ret >= 0,
+    make_string("Failed to send packet to decoder: ", detail::av_error_string(ret)));
+  flush_state_ = true;
+
+  return false;
+}
+
+void VideoFileCPU::Reset() {
+  int ret = av_seek_frame(av_state_->ctx_, av_state_->stream_id_, 0, AVSEEK_FLAG_FRAME);
+  DALI_ENFORCE(
+    ret >= 0,
+    make_string(
+      "Could not seek to the first frame of video ",
+      filename_,
+      "due to",
+      detail::av_error_string(ret)));
+  avcodec_flush_buffers(av_state_->codec_ctx_);
+}
+
+void VideoFileCPU::SeekFrame(int frame_id) {
+  // TODO(awolant): Optimize seeking: 
+  //  - when we seek next frame,
+  //  - when we seek frame with the same keyframe as the current frame
+  //  - for CFR, when we know pts, but don't know keyframes 
+
+  DALI_ENFORCE(
+    frame_id >= 0 && frame_id < NumFrames(),
+    make_string("Invalid seek frame id. frame_id = ", frame_id, ", num_frames = ", NumFrames()));
+
+  auto &frame_entry = index_[frame_id];
+  int keyframe_id = frame_entry.last_keyframe_id;
+  auto &keyframe_entry = index_[keyframe_id];
+
+  // Seeking clears av buffers, so reset flush state info
```

Comment:
Is the av buffer attached to the given file context or is it global?
I just wonder what happens when we interleave decoding of multiple files.

Reply:
As far as I can tell, when we call `av_flush_buffers` it flushes all the buffers in the given `AVFormatContext` and resets. Since we use separate contexts for separate files, nothing gets lost.
Btw. we also still hold any frames we have referenced, as `libav` knows this and keeps them as long as we hold the references.